File: clients/android/app/src/main/java/local/visual/corepose/PoseTimeline.kt
package local.visual.corepose

import kotlin.math.abs

data class PoseFrame(
    val tMs: Long,
    val poses: List<CocoPose>,
    val width: Int,
    val height: Int,
    val blaze33: List<BlazeJoint> = emptyList(),
    val bbox: BBox? = null,
)

object PoseTimeline {
    fun nearest(frames: List<PoseFrame>, tMs: Long): PoseFrame? {
        if (frames.isEmpty()) {
            return null
        }
        val i = frames.binarySearchBy(tMs) { it.tMs }
        if (i >= 0) {
            return frames[i]
        }
        val insert = -i - 1
        val lo = (insert - 1).coerceAtLeast(0)
        val hi = insert.coerceAtMost(frames.lastIndex)
        return if (abs(frames[lo].tMs - tMs) <= abs(frames[hi].tMs - tMs)) {
            frames[lo]
        } else {
            frames[hi]
        }
    }
}

File: clients/ios/VisualPose/PoseTimeline.swift
import Foundation

struct PoseFrame {
    var tMs: Int64
    var poses: [CocoPose]
    var width: Int
    var height: Int
    var blaze33: [BlazeJoint] = []
    var bbox: BBox? = nil
}

enum PoseTimeline {
    static func nearest(_ frames: [PoseFrame], tMs: Int64) -> PoseFrame? {
        if frames.isEmpty { return nil }
        var lo = 0
        var hi = frames.count
        while lo < hi {
            let mid = lo + (hi - lo) / 2
            if frames[mid].tMs < tMs {
                lo = mid + 1
            } else {
                hi = mid
            }
        }
        if lo < frames.count && frames[lo].tMs == tMs {
            return frames[lo]
        }
        let insert = lo
        let loI = max(insert - 1, 0)
        let hiI = min(insert, frames.count - 1)
        if abs(frames[loI].tMs - tMs) <= abs(frames[hiI].tMs - tMs) {
            return frames[loI]
        }
        return frames[hiI]
    }
}
